Add blog post about cpts and taxonomies (#208)
infinum/eightshift-docs: website/blog/2022-12-13-using-cpts-and-taxonomies.md

---
title: Using Custom Post Types and Taxonomies
description: Explains how to register and use custom post types and taxonomies
slug: using-cpts-and-taxonomies
authors: obradovic
date: 2022-12-13
tags: [eightshift, boilerplate, cpt, custom post type, taxonomy, taxonomies, terms]
hide_table_of_contents: false
---

> By default, WordPress offers two post types for our content management - Posts and Pages. But sometimes, that is not enough. That's where custom post types and custom taxonomies come in. This blog post will cover the basics of registering CPTs and Taxonomies with Eightshift Dev kit.

Author: I wrote that a bit wrong 😅  I had that which you wrote in mind, but I haven't explained it well enough I guess

```bash
wp boilerplate create post-type --label='Project' --plural_label='Projects' --slug='project' --rewrite_url='project' --rest_endpoint_slug='projects'
```

> If your new CPT is not working, try flushing rewrite rules by re-saving the settings in **Settings -> Permalinks** or by using `wp cache flush` CLI command

Your new post type is registered and ready to use! Easy, right?

Sometimes all these parameters can be a bit confusing, so here's a quick reference of best practices when setting these parameters:

| Parameter          | Singular/Plural | Writing style | Example  |
|--------------------|-----------------|---------------|----------|
| label              | Singular        | Regular       | Project  |
| plural_label       | Plural          | Regular       | Projects |
| slug               | Singular        | kebab-case    | project  |
| rewrite_url        | Singular        | kebab-case    | project  |
| rest_endpoint_slug | Plural          | kebab-case    | projects |

> 

Reviewer: We could add some explanation why this is the convention.

For instance, the slug will be attached to a _single_ custom post type in the database, which is why it should be a singular lower kebab-cased string.

The rest endpoint slug will expose a _collection_ of custom post types, which is why we want that to be a plural kebab-cased string.

## Registering Taxonomies

> Now that we have the new custom post type, we need a way to group the projects. We'll create a custom taxonomy called `Project Category`. As with the CPT registration, the easiest way to register taxonomies is by using the following WP-CLI command:

Reviewer: `Project Category` is a bit lazy.

Can you think of a more real-life categorization of your projects? 😄

Author: It's not lazy, it's general-purpose 😁 Some other alternatives would be `Project Type` or `Project Scope`. These two are the only ones that come to mind currently 😅

Reviewer 2: Maybe `Project Technology` or `Service`, and group them similarly to how an agency website would (e.g. terms`WordPress`, `iOS`, `Android` for technologies, or `Development`, `Design`, `Project management` for services?)

Reviewer: > It's not lazy, it's general-purpose

Nice one xD

I'd use the suggestions Borna wrote 😄

Author: Updated to `Project Technology` 🙂

```bash
wp boilerplate create taxonomy --label='Project Category' --plural_label='Project Categories' --slug='project-category' --rest_endpoint_slug='project-categories' --post_type_slug='project'
```

Similar suggestions apply to the parameters when naming taxonomies as well. Be sure to write the correct post type slug for which you are registering this new taxonomy!

> 

Reviewer: I'd also add here a bit about why the naming is better, and how taxonomy and terms are organized in the DB.

Mentioning many to many relationships, etc.

Author: I added that part on the bottom, under the `Further Reading` section. WP docs have it explained quite nicely.

## Structure

> When checking your codebase after adding these new custom post types and taxonomies, you'll notice the post types are located inside `src/CustomPostType` folder, and the taxonomies are located inside `src/CustomTaxonomy` folder. Following the **Single Responsibility Principle**, each post type or taxonomy is a separate class.

Reviewer: ```suggestion
When checking your codebase after adding these new custom post types and taxonomies, you'll notice the post types are located inside the `src/CustomPostType` folder, and the taxonomies are located inside the `src/CustomTaxonomy` folder. Following the **Single Responsibility Principle**, each post type or taxonomy is in a separate class.
```

## Modifying options

> Our custom post type and taxonomy are ready to use, but we still want to make some changes. For starters, we want another icon and for the Projects to be located below the Posts in the admin menu. In `src/CustomPostType/ProjectPostType.php`, find the `MENU_POSITION` constant and change it to `5`. The lower the number, the higher it will be in the menu.

Reviewer 3: ```suggestion
Our custom post type and taxonomy are ready to use, but we still want to make some changes. For starters, we want another icon and for the Projects to be located below the Posts in WordPress admin menu. In `src/CustomPostType/ProjectPostType.php`, find the `MENU_POSITION` constant and change it to `5`. The lower the number, the higher it will be in the menu.
```

Reviewer 3: As a first reference to the menu it might be good to say explicitly which menu it is :)

> Next, we want to update the dashicon. Change the `MENU_ICON` constant to `dashicons-clipboard`. If you want another icon for your CPT, here is the list of [available dashicons](https://developer.wordpress.org/resource/dashicons/). Finally, we want to remove the author and comments. In `getPostTypeArguments()` method, find the key `supports` in the return value and remove `author` and `comments` from the array.

Reviewer 3: ```suggestion
Next, we want to update the icon representing the new post type in the menu. These icons are named dashicons. Change the `MENU_ICON` constant to `dashicons-clipboard`. If you want another icon for your CPT, here is the list of [available dashicons](https://developer.wordpress.org/resource/dashicons/). Finally, we want to remove the author and comments. In `getPostTypeArguments()` method, find the key `supports` in the return value and remove `author` and `comments` from the array.
```
